src/System.Net.Security/src/System/Net/CertificateValidationPal.OSX.cs

```diff
@@ -0,0 +1,167 @@
+            return GetRemoteCertificate(securityContext, null);
+        }
+
+        internal static X509Certificate2 GetRemoteCertificate(
+            SafeDeleteContext securityContext,
+            out X509Certificate2Collection remoteCertificateStore)
+        {
+            if (securityContext == null)
+            {
+                remoteCertificateStore = null;
+                return null;
+            }
+
+            remoteCertificateStore = new X509Certificate2Collection();
+            return GetRemoteCertificate(securityContext, remoteCertificateStore);
+        }
+
+        private static X509Certificate2 GetRemoteCertificate(
+            SafeDeleteContext securityContext,
+            X509Certificate2Collection remoteCertificateStore)
+        {
+            if (securityContext == null)
+            {
+                return null;
+            }
+
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(securityContext);
+
+            SafeSslHandle sslContext = ((SafeDeleteSslContext)securityContext).SslContext;
+
+            if (sslContext == null)
+            {
+                return null;
+            }
+
+            X509Certificate2 result = null;
+
+            using (SafeX509ChainHandle chainHandle = Interop.AppleCrypto.SslCopyCertChain(sslContext))
+            {
+                long chainSize = Interop.AppleCrypto.X509ChainGetChainSize(chainHandle);
+
+                if (remoteCertificateStore != null)
+                {
+                    for (int i = 0; i < chainSize; i++)
+                    {
+                        IntPtr certHandle = Interop.AppleCrypto.X509ChainGetCertificateAtIndex(chainHandle, i);
+                        remoteCertificateStore.Add(new X509Certificate2(certHandle));
+                    }
+                }
+
+                // This will be a distinct object than remoteCertificateStore[0] (if applicable),
+                // to match what the Windows and Unix PALs do.
+                if (chainSize > 0)
+                {
+                    IntPtr certHandle = Interop.AppleCrypto.X509ChainGetCertificateAtIndex(chainHandle, 0);
+                    result = new X509Certificate2(certHandle);
+                }
+            }
+
+            if (NetEventSource.IsEnabled)
+            {
+                NetEventSource.Log.RemoteCertificate(result);
+                NetEventSource.Exit(securityContext, result);
+            }
+
+            return result;
+        }
+
+        //
+        // Used only by client SSL code, never returns null.
+        //
+        internal static string[] GetRequestCertificateAuthorities(SafeDeleteContext securityContext)
+        {
+            SafeSslHandle sslContext = ((SafeDeleteSslContext)securityContext).SslContext;
+
+            if (sslContext == null)
+            {
+                return Array.Empty<string>();
+            }
+
+            using (SafeCFArrayHandle dnArray = Interop.AppleCrypto.SslCopyCADistinguishedNames(sslContext))
```

Comment:
By default, a CFArray CFRetain()s (AddRef) items added to it, and CFRelease()s (DownRef) those items when they are removed (including the array itself being CFRelease()d for the last time).

Further, OSX has a Get vs Create/Copy convention.  If you "Get" an item, it has not been CFRetain()ed by the caller.  If you Copy it, it has (therefore it's yours to CFRelease).

So, in this method we "Copy" the names (we own the array, all the items were up-reffed when the array was created to track that the array knows about them), then we "Get" the values (so we don't own them), we then extract the data we need, and don't (because ownsHandle is false) CFRelease the individal members.  We do, at the end, CFRelease (via ReleaseHandle) the array, which down-refs all of the data elements.

If SecureTransport gave us shared elements they'll still be alive.  If they were uniquely constructed for this return they'll get cleaned up by CFRelease.

Most of that is conveyed in Apple documentation by "follows [The Create Rule](https://developer.apple.com/library/content/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html)"